Lobby Refactor

Jezithyr opened the pull request:
I rewrote the lobby ui and implemented functionality for custom background images for lobbies. Also I cleaned up alot of the backend ui scripting for the lobby, it should be less spaghetti now. 

Additionally I created some controls for common ui formating, HLine, HSpacer, and VSpacer. They do exactly what they say they do. This helps cut down on the amount of copy pasting in XAML Files.

There is now a system for custom lobby screen images, simply drop the images into the /Server Info/LobbyBackgrounds and the server will randomly select one to use each lobby. To disable an image simply rename it's file extension.

I also added an AdminHelp button to the lobby screen which fixes #6987. Now you can complain to badmins before the game even starts!

Requires: https://github.com/space-wizards/RobustToolbox/pull/2599


### WEH
![image](https://user-images.githubusercontent.com/6192499/158012682-39c01215-a2e2-4cea-97dc-5f7cdcb59593.png)


:cl: Jezithyr
- add: Rewrote the lobby UI
- add: Added adminhelp button to the lobby screen
- add: Added the ability to add custom lobby screens, (You can have multiple and picks a random one!)
- tweak: rewrote the backend of the lobby ui to be slightly less ass, still a work in progress. Please report any issues with monitor scaling.

metalgearsloth commented:
![image](https://user-images.githubusercontent.com/31366439/158017628-66c272e5-fd71-4950-9021-a9f9fda1a7b6.png)

metalgearsloth reviewed (approved):
Menu is a bit narrower than I'd like for chat purposes, rest seems good.

![image](https://user-images.githubusercontent.com/31366439/158017854-62761c51-d6f8-42b3-b211-7bbb93f61461.png)

ElectroJr commented on Content.Client/Lobby/LobbyState.cs:
```diff
@@ -222,37 +225,52 @@ private void UpdateLobbyUi()
         }
 
+
+        //private void UpdatePlayerList()
+        //{
+        //    if (_lobby == null) return;
+        //    _lobby.OnlinePlayerList.Clear();
+        //    var gameTicker = EntitySystem.Get<ClientGameTicker>();
+
+        //    foreach (var session in _playerManager.Sessions.OrderBy(s => s.Name))
+        //    {
+        //        var readyState = string.Empty;
+        //        // Don't show ready state if we're ingame
+        //        if (!gameTicker.IsGameStarted)
+        //        {
+        //            LobbyPlayerStatus status;
+        //            if (session.UserId == _playerManager.LocalPlayer?.UserId)
+        //                status = gameTicker.AreWeReady ? LobbyPlayerStatus.Ready : LobbyPlayerStatus.NotReady;
+        //            else
+        //                gameTicker.Status.TryGetValue(session.UserId, out status);
+
+        //            readyState = status switch
+        //            {
+        //                LobbyPlayerStatus.NotReady => Loc.GetString("lobby-state-player-status-not-ready"),
+        //                LobbyPlayerStatus.Ready => Loc.GetString("lobby-state-player-status-ready"),
+        //                LobbyPlayerStatus.Observer => Loc.GetString("lobby-state-player-status-observer"),
+        //                _ => string.Empty,
+        //            };
+        //        }
+
+        //        _lobby.OnlinePlayerList.AddItem(session.Name, readyState);
+        //    }
+        //}
```
Is the plan add the player list back in? Otherwise why is all the code here and throughout `LobbyState.cs` just commented out?

Jezithyr replied:
Woops totally forgot to clean that up, good catch. No plans to re add the lobby playerlist, I was worried about dependencies so i commented it out for testing and forgot about it

ElectroJr reviewed (commented):
There seems to be some layout issue with the text chat, where AFAICT it goes above the separator. Also for smaller windows (screenshot is ~720p window), the "main" window is probably too wide? It seems to auto-resize when the round starts, I'm guessing it just expands to fit all of the yellow text? If that's the case that should probably be changed in case anyone ever writes an even longer game mode description.

![text-escape](https://user-images.githubusercontent.com/60421075/158019272-67163890-b9fc-4ea0-ac4d-20df39e6f962.png)

Jezithyr commented on Content.Client/Lobby/UI/LobbyCharacterPreviewPanel.cs:
```diff
@@ -32,13 +32,12 @@ public sealed class LobbyCharacterPreviewPanel : Control
-            _entMan = entityManager;
-            _preferencesManager = preferencesManager;
-            _prototypeManager = prototypeManager;
+            //This only runs once, if you cry about optimization fite me - Jezi
+            _entMan = IoCManager.Resolve<IEntityManager>();
+            _preferencesManager = IoCManager.Resolve<IClientPreferencesManager>();
+            _prototypeManager = IoCManager.Resolve<IPrototypeManager>();
```
What do you recommend then? [Injecting] them? I can convert it over to use that method. The constructor needs to have no arguments otherwise it causes issues with the XAML. I'll look into migrating all that away from ioc manager

metalgearsloth replied:
acruid had this to say: https://discord.com/channels/310555209753690112/900426319433728030/952346212470759424

Jezithyr commented:
> There seems to be some layout issue with the text chat, where AFAICT it goes above the separator. Also for smaller windows (screenshot is ~720p window), the "main" window is probably too wide? It seems to auto-resize when the round starts, I'm guessing it just expands to fit all of the yellow text? If that's the case that should probably be changed in case anyone ever writes an even longer game mode description.
> 
> ![text-escape](https://user-images.githubusercontent.com/60421075/158019272-67163890-b9fc-4ea0-ac4d-20df39e6f962.png)

I think that's an issue with the chat control scaling, when it gets crushed the text escapes the control's container for some reason. This will be fixed while I'm working on refactoring chat/oldchat.

Jezithyr commented:
Fixed chat in 720p. Unfortunately 4:3 is still going to cause issues without implementing some sort of autoscale system for UI
![image](https://user-images.githubusercontent.com/6192499/158036720-14ad7a17-b346-459d-aabf-53ae4d454a89.png)

moonheart08 commented:
> Fixed chat in 720p. Unfortunately 4:3 is still going to cause issues without implementing some sort of autoscale system for UI ![image](https://user-images.githubusercontent.com/6192499/158036720-14ad7a17-b346-459d-aabf-53ae4d454a89.png)

We have players on 4:3 so this needs to work for this PR to be approved. Sorry.

mirrorcult commented on Content.Client/UserInterface/Controls/HLine.cs:
```diff
@@ -0,0 +1,57 @@
+    }
+
+}
+
+
+/*
+ <PanelContainer>
+                    <PanelContainer.PanelOverride>
+                        <gfx:StyleBoxFlat BackgroundColor="{x:Static style:StyleNano.NanoGold}"
+                                          ContentMarginTopOverride="2" />
+                    </PanelContainer.PanelOverride>
+                </PanelContainer>
+
+*/
```
remove

mirrorcult commented on Content.Server/GameTicking/GameTicker.LobbyBackground.cs:
```diff
@@ -0,0 +1,33 @@
+
+    private static readonly string[] WhitelistedBackgroundExtensions = new string[] {"png", "jpg", "jpeg"};
+
+    private const string LobbyScreenPath = "/Server Info/LobbyScreens";
```
`Server Info` is a bad path for this, it should be in `Textures`

Jezithyr replied:
Yep good call, move to textures. I thought it made more sense to be in server info but that was at 1am :P

mirrorcult reviewed (changes requested):
might as well nitpick since its blocked on 4:3 anyway

PaulRitter commented on Content.Client/Lobby/UI/LobbyCharacterPreviewPanel.cs:
```diff
@@ -58,7 +55,11 @@ public sealed class LobbyCharacterPreviewPanel : Control
             };
 
-            vBox.AddChild(header);
+            var vSpacer = new BoxContainer
+            {
+                Orientation = LayoutOrientation.Vertical,
+                MinHeight = 5
+            };
```
use the type you created for this

Jezithyr replied:
I missed this one, good catch. I'm pushing a fix

PaulRitter reviewed (changes requested):
a pr so good it brought me out of hibernation

lgtm besides the 4:3 thing & the nitpick i found

PaulRitter commented:
now do oldchat :^)

DogZeroX commented:
They are already doing it, Paul.

Jezithyr commented:
> now do oldchat :^)

:^3

Jezithyr commented:
Fix for 4:3 requires: https://github.com/space-wizards/RobustToolbox/pull/2599

Jezithyr commented:
Updated the layout a bit, the link button panel was getting a bit too crowded and was causing overlap issues on low resolutions.
I moved the Report bugs and credits buttons to the bottom left to prevent them from overlapping the other buttons on low resolutions.
![Content Client_4RcgXOru1b](https://user-images.githubusercontent.com/6192499/158081846-ac486e26-6e28-46a6-ad1c-f13d35a89341.png)
Honestly, I also like the look. It clearly separates the server specific info (rules, discord, changelog, wiki) from the more developer info stuff (report bugs, credits etc.). It also helps emphasize the Report Bugs button which hopefully should result in users being more likely to report bugs.

BozoDel commented:
It still overlaps for me sometimes, at 1024x768 resolution.

But eh, so far I haven't seen it completely cover the Ready Up button, so I don't mind.
